R/gen_population.R

```diff
@@ -6,66 +6,128 @@
+  nmeasure <- length(measure)
 
+  # define a dataframe based on groups only
+  output <- data.frame(groups)
 
+  # if strata specified then make a dataframe from combining groups*strata
   if (!is.null(strata)) {
 
-    strata2 <- factor(
-                      rep.int(strata, length(groups)),
-                      levels = strata
-                      )
-    strata2 <- sort(strata2)
+    # give error - check length of given input measure and groups match with strata
+    if (nmeasure != ngroups && nmeasure != ngroups * nstrata) {
 
+      differences <- abs(ngroups - nmeasure)
+      differences2 <- abs(ngroups * nstrata - nmeasure)
 
-    groups2 <-  factor(
-                       rep.int(groups, length(strata)),
-                       levels = groups
-                       )
+      stop(glue::glue("Given proportions (or counts) and groups lengths",
+                      "do not match, nor are they a multiple of strata\n",
+                      "The difference in length was {differences}\n",
+                      "The difference in multiplied length was {differences2}"))
+    }
 
-    if (length(proportions) < length(groups2)) {
-      proportions2 <- rep.int(proportions, length(strata))
+    # give warning if repeating input measures for strata
+    if (nmeasure != ngroups * nstrata) {
+      warning(glue::glue("Given proportions (or counts) is not the same as\n",
+                         "groups multiplied by strata length, they will be repeated to match"))
     }
 
-    output <- bind_cols(groups = groups2,
-                        strata = strata2,
-                        proportions = proportions2,
-                        n = proportions2 * total_pop)
+    # create data frame with groups and strata
+    output <- expand.grid(groups = groups, strata = strata)
+  } else {
+    # check length of given input measure and groups match without strata
+    if (nmeasure != ngroups) {
+      differences <- abs(ngroups - nmeasure)
+
+      stop(glue::glue("Given proportions (or counts) and groups lengths
+                    do not match and there were no strata\n",
+                      "The difference in length was {differences}"))
+    }
+  }
+
+
+  # for proportions
+  if (is.null(counts)) {
+    # pull together the props
+    output$proportions <- proportions
+
+    # create counts based on multiplying by total population
+    output$n <- output$proportions * total_pop
+
+    # give warning if proportions are more than 5% away from 100%
+    if (sum(output$proportions) > 1.05 | sum(output$proportions) < 0.95) {
+      warning(glue::glue("The sum of given proportions is more than 5% away from 100% \n",
+                         "If the intention was to provide proportions by strata then ignore this message"))
+    }
 
   } else {
-    output <- bind_cols(groups = groups,
-                        proportions = proportions,
-                        n = proportions * total_pop)
+    # for counts
+    output$n <- counts
```

Comment:
```suggestion
output$proportions <- counts / sum(counts)
output$n <- counts
```

Reply:
once this is implemented, then the order will be correct

Reply:
cant do this way round because have to create n first - because creating n is what repeats counts for stratified. (if just sum counts as in this suggestion then you get the wrong number!)

Reply:
good point!